plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogAccess.java

```diff
@@ -190,6 +204,83 @@ public TableSnapshot loadSnapshot(ConnectorSession session, SchemaTableName tabl
         return snapshot;
     }
 
+    private TableSnapshot loadSnapshotForTimeTravel(TrinoFileSystem fileSystem, SchemaTableName table, String tableLocation, long endVersion)
+            throws IOException
+    {
+        try {
+            return TableSnapshot.load(
+                    table,
+                    findCheckpoint(fileSystem, tableLocation, endVersion),
+                    fileSystem,
+                    tableLocation,
+                    parquetReaderOptions,
+                    checkpointRowStatisticsWritingEnabled,
+                    domainCompactionThreshold,
+                    Optional.of(endVersion));
+        }
+        catch (UncheckedExecutionException e) {
+            throwIfUnchecked(e.getCause());
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static Optional<LastCheckpoint> findCheckpoint(TrinoFileSystem fileSystem, String tableLocation, long endVersion)
+    {
+        Optional<LastCheckpoint> lastCheckpoint = readLastCheckpoint(fileSystem, tableLocation);
+        if (lastCheckpoint.isPresent() && lastCheckpoint.get().version() <= endVersion) {
+            return lastCheckpoint;
+        }
+
+        // TODO https://github.com/trinodb/trino/issues/21366 Make this logic efficient
+        Optional<LastCheckpoint> latestCheckpoint = Optional.empty();
+        Location transactionDirectory = Location.of(getTransactionLogDir(tableLocation));
+        try {
+            FileIterator files = fileSystem.listFiles(transactionDirectory);
+            while (files.hasNext()) {
+                FileEntry file = files.next();
+                Optional<LastCheckpoint> checkpoint = extractCheckpointVersion(file);
+                if (checkpoint.isEmpty()) {
+                    continue;
+                }
+
+                long version = checkpoint.get().version();
+                if (version > endVersion || (latestCheckpoint.isPresent() && version < latestCheckpoint.get().version())) {
+                    continue;
+                }
+                latestCheckpoint = checkpoint;
+            }
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return latestCheckpoint;
+    }
+
+    private static Optional<LastCheckpoint> extractCheckpointVersion(FileEntry file)
+    {
+        String fileName = file.location().fileName();
+        Matcher classicCheckpoint = CLASSIC_CHECKPOINT.matcher(fileName);
```

Comment:
can we combine these matchers into a single one? with named groups?

Reply:
I prefer simple separated patterns in this case, but let me know if you still want me to change it.